### DISCUSSION TITLE
Constructors added to structs seem to mess with the serialization.

### DISCUSSION THREAD

**Discussion by CezarZbughin** [stephenberry/glaze#767] [Q&A] [answered]
I am not sure if this is a bug or the glaze contract states that constructors added to struct produce undefined behavior. So I have the following setup.

1. If a struct has a constructor (No Args constructor) we get a compile time assert that says the type requires metadata.
2. If the struct's metadata is defined then it works fine.
3. If there is another constructor (other than the No Args consrtuctor) now the structure will serialize fine, unless it is wrapped in another structure; In that case the wrapper will serialize as an empty object.

```
struct A {
    int x = 0;

    A() = default;
    A(int) {}; //Comment & uncomment this constructor.
};

template <>
struct glz::meta<A> {
    using T = A;
    static constexpr auto value = object(&T::x);
};

struct B {
    A a;
};

int main() {
    std::string a_json, b_json;
    B b;
    A a;
    glz::write_json(b, b_json);
    glz::write_json(a, a_json);
    std::cout << "Serialized A: " << a_json << std::endl;
    std::cout << "Serialized B: " << b_json << std::endl;
}
```
Output:
With only one constructor defined:
```
Serialized A: {"x":0}
Serialized B: {"a":{"x":0}}
```

With both constructors defined:
```
Serialized A: {"x":0}
Serialized B: {}
```

Well adding metadata for the B structure fixes the problem, but not having it defined does not assert. It also feels like it should not be necessary to add that metadata since it's only when A is wrapped in another struct that it causes problems.

I did not create an issue since I am not sure if this is a bug or not. I though it is better to just ask here. maybe i am just not understanding the structure of the library good enough or i am doing something wrong.

**Comment by stephenberry** [comment, marked as answer]
I've documented the fix here as well: #721. The solution is to add another constructor that enables brace initialization for Glaze.

Simply add the following constructor: `A(glz::make_reflectable) {}`

In full:
```c++
struct A {
    int x = 0;

    A(glz::make_reflectable) {}
    A() = default;
    A(int) {}; //Comment & uncomment this constructor.
};
```
